### PR TITLE
♻️ `dynamic-sidecar` fails to boot if external dependencies are not present

### PR DESCRIPTION
**Pull request by GitHK**
## What do these changes do?

- ✨ Upon startup the `dynamic-sidecar` makes sure it can connect to `Postgres`, `storage` and `RabbitMQ`. The setup process will fail, if one of them is not reachable.

There is a default window, of `30 seconds`, in which these need to be reachable after which the sidecar will raise an error on startup and the container will fail.

- ♻️ `Posrgres` and `RabbitMQ` settings are no longer optional, but mandatory for the `dinamic-sidecar`. They were already treated like this, so it made no sense to continue having them optional.

## Dev Checklist

- [x] No ENV changes or I properly updated ENV ([read the instruction](https://git.speag.com/oSparc/osparc-ops-deployment-configuration/-/blob/configs/README.md?ref_type=heads#how-to-update-env-variables))

### CODE REVIEW

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/external_dependencies.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+from .postgres import wait_for_postgres_liveness
+from .rabbitmq import wait_for_rabbitmq_liveness
+from .storage import wait_for_storage_liveness
+
+
+def setup_check_dependencies(app: FastAPI) -> None:
+    async def on_startup() -> None:
+        await logged_gather(
+            *[
+                wait_for_storage_liveness(app),
```
Comment (pcrespov) [inline services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/external_dependencies.py:13]: I still think that it would be better to have these in a setup for each service.  i.e. `setup_storage` in `setup_rabbit`, etc

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/service_liveness.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        await _attempt_to_wait_for_handler(
+            async_handler,
+            *args,
+            service_name=service_name,
+            endpoint=endpoint,
+            check_interval=check_interval,
+            timeout=timeout,
+            **kwargs,
+        )
+        elapsed = (time.time() - start) * 1000
```
Comment (pcrespov) [inline services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/service_liveness.py:97]: elapsed_ms

File: services/dynamic-sidecar/tests/integration/test_core_external_dependencies.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        "RABBIT_PASSWORD": rabbit_settings.RABBIT_PASSWORD.get_secret_value(),
+        "RABBIT_PORT": f"{rabbit_settings.RABBIT_PORT}",
+        "RABBIT_SECURE": f"{rabbit_settings.RABBIT_SECURE}",
+        "RABBIT_USER": rabbit_settings.RABBIT_USER,
+        "STORAGE_HOST": storage_endpoint.host,
+        "STORAGE_PORT": f"{storage_endpoint.port}",
+        **base_mock_envs,
+    }
+
+    setenvs_from_dict(monkeypatch, envs)
```
Comment (pcrespov) [inline services/dynamic-sidecar/tests/integration/test_core_external_dependencies.py:57]: MINOR: just

```python
def mock_environments( .. )
   return setenvs_from_dict(monkeypatch,
   {
        "DY_SIDECAR_CALLBACKS_MAPPING": "{}",
        "DY_SIDECAR_PROJECT_ID": f"{project_id}",
        "DY_SIDECAR_USER_ID": f"{user_id}",
        "DYNAMIC_SIDECAR_LOG_LEVEL": "DEBUG",
        "R_CLONE_PROVIDER": "MINIO",
        "RABBIT_HOST": rabbit_settings.RABBIT_HOST,
        "RABBIT_PASSWORD": rabbit_settings.RABBIT_PASSWORD.get_secret_value(),
        "RABBIT_PORT": f"{rabbit_settings.RABBIT_PORT}",
        "RABBIT_SECURE": f"{rabbit_settings.RABBIT_SECURE}",
        "RABBIT_USER": rabbit_settings.RABBIT_USER,
        "STORAGE_HOST": storage_endpoint.host,
        "STORAGE_PORT": f"{storage_endpoint.port}",
        **base_mock_envs,
    }
   )
```

note in addition that the returned dict is not the same as `env`. See the code of `setenvs_from_dict`

File: services/dynamic-sidecar/tests/integration/test_core_external_dependencies.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    mock_registry_service: AsyncMock,
+) -> FastAPI:
+    """creates app with registry and rabbitMQ services mocked"""
+    return create_app()
+
+
+async def test_external_dependencies_are_reachable(app: FastAPI):
+    async with TestClient(app):
+        # checks that client starts properly
+        assert True
```
Comment (pcrespov) [inline services/dynamic-sidecar/tests/integration/test_core_external_dependencies.py:73]: I would check the opposite as well. e.g. stop `storage` and check that it fails to start and how is the system reacting to that failure e.g. who notices? directorv2? is it reported properly etc?

File: services/dynamic-sidecar/tests/unit/test_service_liveness.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
+from datetime import timedelta
+
+import pytest
+from simcore_service_dynamic_sidecar.modules.service_liveness import (
+    CouldNotReachServiceError,
+    wait_for_service_liveness,
+)
+
+
+@pytest.mark.parametrize("check_interval", [timedelta(seconds=0.1)])
```
Comment (pcrespov) [inline services/dynamic-sidecar/tests/unit/test_service_liveness.py:10]: why did you add check_interval and timeout as parameters instead of constant values in the test?

**Review by pcrespov — APPROVED** [review]
thx. Left some suggestions

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/external_dependencies.py (hunk 1 above)
Comment (GitHK) [reply services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/external_dependencies.py:13]: Yeah this one is not obvious why it's like this.
While I agree with you, that it's better to have the checks delegated to their own modules, to achieve a faster start, we run all of them in parallel for the `dy-sidecar`.

File: services/dynamic-sidecar/tests/integration/test_core_external_dependencies.py (hunk 4 above)
Comment (GitHK) [reply services/dynamic-sidecar/tests/integration/test_core_external_dependencies.py:73]: Yes, this makes total sense. You test the failuire.
In this situation:
- `director-v2` waits for the sidecar to start
- `docker` restarts the container if it cannot connect

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/external_dependencies.py (hunk 1 above)
Comment (sanderegg) [reply services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/external_dependencies.py:13]: I don't think it is useful for faster start (the dy-sidecar already takes less than 7 seconds to start).
Maybe in case of multiple dependencies missing... but even there... not sure that makes sense.

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/service_liveness.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    timeout: timedelta,
+    **kwargs,
+) -> None:
+    async for attempt in AsyncRetrying(
+        wait=wait_fixed(check_interval),
+        stop=stop_after_delay(timeout.total_seconds()),
+        before_sleep=_before_sleep_log(_logger, service_name, endpoint),
+        reraise=True,
+    ):
+        with attempt:
```
Comment (sanderegg) [inline services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/service_liveness.py:54]: minor: this form is bad as it does not show in the log which function is retried. better to use the decorator.

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/service_liveness.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+            *args,
+            service_name=service_name,
+            endpoint=endpoint,
+            check_interval=check_interval,
+            timeout=timeout,
+            **kwargs,
+        )
+        elapsed = (time.time() - start) * 1000
+        _logger.info(
+            "Service '%s' found at '%s' after %.2f ms", service_name, endpoint, elapsed
```
Comment (sanderegg) [inline services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/service_liveness.py:99]: I think all that timing stuff is given by tenacity in the retry_state.statistics or so. please search in the code for similar things

File: services/dynamic-sidecar/tests/integration/test_modules_long_running_tasks.py [hunk 8]
```diff
@@ -54,7 +49,7 @@
 
 pytest_simcore_ops_services_selection = [
     "minio",
-    "adminer",
+    # "adminer",
```
Comment (sanderegg) [inline services/dynamic-sidecar/tests/integration/test_modules_long_running_tasks.py:52]: same note as before

**Review by sanderegg — APPROVED** [review]
cool! but please check my comments.

**Review by matusdrobuliak66 — APPROVED** [review]
